File: api/src/services/updates.js
import { exec, spawn } from 'child_process';
import { promisify } from 'util';
import { existsSync } from 'fs';
import { getIO } from '../socket.js';

const execAsync = promisify(exec);

// State management
let activeCheckProcess = null;
let checkCancelled = false;
let lastCheckResult = null;

// Check if a check is running
export function isCheckRunning() {
  return activeCheckProcess !== null;
}

// Get last check result
export function getLastCheckResult() {
  return lastCheckResult;
}

// Cancel running check
export async function cancelCheck() {
  if (!activeCheckProcess) {
    return { success: false, error: 'No check in progress' };
  }

  checkCancelled = true;

  try {
    activeCheckProcess.kill('SIGTERM');
  } catch {
    // Process may have already exited
  }

  getIO().emit('updates:cancelled', {});

  return { success: true, message: 'Check cancellation requested' };
}

// Parse apt list --upgradable output
function parseAptList(output) {
  const lines = output.trim().split('\n').filter(l => l.includes('[upgradable'));
  return lines.map(line => {
    // Format: package/repo version arch [upgradable from: oldversion]
    const match = line.match(/^([^/]+)\/([^\s]+)\s+([^\s]+)\s+([^\s]+)\s+\[upgradable from: ([^\]]+)\]/);
    if (!match) return null;
    const [, name, repo, newVersion, arch, currentVersion] = match;
    return {
      name,
      repository: repo,
      isSecurity: repo.toLowerCase().includes('security'),
      currentVersion,
      newVersion,
      arch
    };
  }).filter(Boolean);
}

// Parse snap refresh --list output
function parseSnapList(output) {
  const lines = output.trim().split('\n');
  if (lines.length <= 1) return []; // Only header or empty

  // Skip header line
  return lines.slice(1).map(line => {
    const parts = line.trim().split(/\s+/);
    if (parts.length < 4) return null;
    return {
      name: parts[0],
      newVersion: parts[1],
      revision: parts[2],
      publisher: parts[3]
    };
  }).filter(Boolean);
}

// Parse needrestart -b output
function parseNeedrestart(output) {
  const result = {
    kernelRebootNeeded: false,
    currentKernel: null,
    expectedKernel: null,
    services: []
  };

  const lines = output.split('\n');
  for (const line of lines) {
    if (line.startsWith('NEEDRESTART-KSTA:')) {
      const status = parseInt(line.split(':')[1].trim());
      result.kernelRebootNeeded = status > 1; // 1=current, 2=ABI compatible, 3=different
    } else if (line.startsWith('NEEDRESTART-KCUR:')) {
      result.currentKernel = line.split(':')[1].trim();
    } else if (line.startsWith('NEEDRESTART-KEXP:')) {
      result.expectedKernel = line.split(':')[1].trim();
    } else if (line.startsWith('NEEDRESTART-SVC:')) {
      result.services.push(line.split(':')[1].trim());
    }
  }

  return result;
}

// Run apt update with streaming output
function runAptUpdate() {
  return new Promise((resolve, reject) => {
    const io = getIO();
    const apt = spawn('sudo', ['apt', 'update']);
    activeCheckProcess = apt;

    let stdout = '';
    let stderr = '';

    apt.stdout.on('data', (data) => {
      const chunk = data.toString();
      stdout += chunk;
      // Stream each line
      const lines = chunk.split('\n').filter(Boolean);
      for (const line of lines) {
        io.emit('updates:output', { phase: 'apt-update', line });
      }
    });

    apt.stderr.on('data', (data) => {
      stderr += data.toString();
    });

    apt.on('close', (code) => {
      activeCheckProcess = null;
      if (checkCancelled) {
        reject(new Error('Check cancelled'));
      } else if (code === 0) {
        resolve({ stdout, stderr });
      } else {
        reject(new Error(`apt update failed with code ${code}: ${stderr}`));
      }
    });

    apt.on('error', (err) => {
      activeCheckProcess = null;
      reject(err);
    });
  });
}

// Main check function
export async function runFullCheck() {
  if (activeCheckProcess) {
    return { success: false, error: 'A check is already in progress' };
  }

  const io = getIO();
  const startTime = Date.now();
  checkCancelled = false;

  const result = {
    timestamp: new Date().toISOString(),
    duration: 0,
    apt: { packages: [], securityCount: 0, normalCount: 0 },
    snap: { packages: [] },
    needrestart: { services: [], kernelRebootNeeded: false },
    summary: { totalUpdates: 0, securityUpdates: 0, servicesNeedingRestart: 0 }
  };

  try {
    io.emit('updates:started', { timestamp: result.timestamp });

    // Phase 1: APT update
    io.emit('updates:phase', { phase: 'apt-update', message: 'Mise a jour des listes de paquets...' });
    try {
      await runAptUpdate();
    } catch (err) {
      if (checkCancelled) throw err;
      console.error('apt update error:', err.message);
    }

    if (checkCancelled) throw new Error('Check cancelled');

    // Phase 2: APT list upgradable
    io.emit('updates:phase', { phase: 'apt-list', message: 'Verification des paquets...' });
    try {
      const { stdout } = await execAsync('apt list --upgradable 2>/dev/null', { timeout: 60000 });
      result.apt.packages = parseAptList(stdout);
      result.apt.securityCount = result.apt.packages.filter(p => p.isSecurity).length;
      result.apt.normalCount = result.apt.packages.length - result.apt.securityCount;
      io.emit('updates:apt-complete', {
        packages: result.apt.packages,
        securityCount: result.apt.securityCount,
        normalCount: result.apt.normalCount
      });
    } catch (err) {
      console.error('apt list error:', err.message);
    }

    if (checkCancelled) throw new Error('Check cancelled');

    // Phase 3: Snap refresh list
    io.emit('updates:phase', { phase: 'snap', message: 'Verification des snaps...' });
    try {
      const { stdout } = await execAsync('snap refresh --list 2>/dev/null', { timeout: 60000 });
      result.snap.packages = parseSnapList(stdout);
      io.emit('updates:snap-complete', { snaps: result.snap.packages });
    } catch (err) {
      // snap refresh --list returns non-zero if no updates
      if (!err.message.includes('All snaps up to date')) {
        console.error('snap refresh error:', err.message);
      }
    }

    if (checkCancelled) throw new Error('Check cancelled');

    // Phase 4: Needrestart
    io.emit('updates:phase', { phase: 'needrestart', message: 'Detection des services a redemarrer...' });
    try {
      // Check if needrestart is available
      const { stdout: whichOutput } = await execAsync('which needrestart 2>/dev/null');
      if (whichOutput.trim()) {
        const { stdout } = await execAsync('sudo needrestart -b 2>/dev/null', { timeout: 60000 });
        result.needrestart = parseNeedrestart(stdout);
      } else {
        // Fallback: check reboot-required file
        if (existsSync('/var/run/reboot-required')) {
          result.needrestart.kernelRebootNeeded = true;
        }
        // Try to get services from /var/run/reboot-required.pkgs
        if (existsSync('/var/run/reboot-required.pkgs')) {
          try {
            const { stdout } = await execAsync('cat /var/run/reboot-required.pkgs');
            result.needrestart.services = stdout.trim().split('\n').filter(Boolean);
          } catch {}
        }
      }
      io.emit('updates:needrestart-complete', result.needrestart);
    } catch (err) {
      console.error('needrestart error:', err.message);
    }

    // Calculate summary
    result.summary = {
      totalUpdates: result.apt.packages.length + result.snap.packages.length,
      securityUpdates: result.apt.securityCount,
      servicesNeedingRestart: result.needrestart.services.length,
      kernelRebootNeeded: result.needrestart.kernelRebootNeeded
    };

    result.duration = Date.now() - startTime;
    lastCheckResult = result;

    io.emit('updates:complete', {
      success: true,
      summary: result.summary,
      duration: result.duration
    });

    return { success: true, result };

  } catch (err) {
    const duration = Date.now() - startTime;

    if (checkCancelled) {
      io.emit('updates:cancelled', {});
      return { success: false, error: 'Check cancelled', cancelled: true };
    }

    io.emit('updates:error', { error: err.message });
    return { success: false, error: err.message };
  } finally {
    activeCheckProcess = null;
  }
}

// State for upgrade operations
let activeUpgradeProcess = null;
let upgradeCancelled = false;

export function isUpgradeRunning() {
  return activeUpgradeProcess !== null;
}

export async function cancelUpgrade() {
  if (!activeUpgradeProcess) {
    return { success: false, error: 'No upgrade in progress' };
  }

  upgradeCancelled = true;

  try {
    activeUpgradeProcess.kill('SIGTERM');
  } catch {
    // Process may have already exited
  }

  getIO().emit('updates:upgrade-cancelled', {});

  return { success: true, message: 'Upgrade cancellation requested' };
}

// Run apt upgrade with streaming output
export async function runAptUpgrade(fullUpgrade = false) {
  if (activeUpgradeProcess) {
    return { success: false, error: 'An upgrade is already in progress' };
  }

  const io = getIO();
  const startTime = Date.now();
  upgradeCancelled = false;

  const command = fullUpgrade ? 'full-upgrade' : 'upgrade';

  return new Promise((resolve) => {
    io.emit('updates:upgrade-started', { type: 'apt', command });

    const apt = spawn('sudo', ['apt', command, '-y'], {
      env: { ...process.env, DEBIAN_FRONTEND: 'noninteractive' }
    });
    activeUpgradeProcess = apt;

    let stdout = '';
    let stderr = '';

    apt.stdout.on('data', (data) => {
      const chunk = data.toString();
      stdout += chunk;
      const lines = chunk.split('\n').filter(Boolean);
      for (const line of lines) {
        io.emit('updates:upgrade-output', { type: 'apt', line });
      }
    });

    apt.stderr.on('data', (data) => {
      const chunk = data.toString();
      stderr += chunk;
      const lines = chunk.split('\n').filter(Boolean);
      for (const line of lines) {
        io.emit('updates:upgrade-output', { type: 'apt', line });
      }
    });

    apt.on('close', (code) => {
      activeUpgradeProcess = null;
      const duration = Date.now() - startTime;

      if (upgradeCancelled) {
        io.emit('updates:upgrade-cancelled', { type: 'apt' });
        resolve({ success: false, error: 'Upgrade cancelled', cancelled: true });
      } else if (code === 0) {
        io.emit('updates:upgrade-complete', { type: 'apt', success: true, duration });
        resolve({ success: true, duration });
      } else {
        io.emit('updates:upgrade-complete', { type: 'apt', success: false, error: stderr });
        resolve({ success: false, error: `apt ${command} failed with code ${code}` });
      }
    });

    apt.on('error', (err) => {
      activeUpgradeProcess = null;
      io.emit('updates:upgrade-complete', { type: 'apt', success: false, error: err.message });
      resolve({ success: false, error: err.message });
    });
  });
}

// Run snap refresh with streaming output
export async function runSnapRefresh() {
  if (activeUpgradeProcess) {
    return { success: false, error: 'An upgrade is already in progress' };
  }

  const io = getIO();
  const startTime = Date.now();
  upgradeCancelled = false;

  return new Promise((resolve) => {
    io.emit('updates:upgrade-started', { type: 'snap' });

    const snap = spawn('sudo', ['snap', 'refresh']);
    activeUpgradeProcess = snap;

    let stdout = '';
    let stderr = '';

    snap.stdout.on('data', (data) => {
      const chunk = data.toString();
      stdout += chunk;
      const lines = chunk.split('\n').filter(Boolean);
      for (const line of lines) {
        io.emit('updates:upgrade-output', { type: 'snap', line });
      }
    });

    snap.stderr.on('data', (data) => {
      const chunk = data.toString();
      stderr += chunk;
      const lines = chunk.split('\n').filter(Boolean);
      for (const line of lines) {
        io.emit('updates:upgrade-output', { type: 'snap', line });
      }
    });

    snap.on('close', (code) => {
      activeUpgradeProcess = null;
      const duration = Date.now() - startTime;

      if (upgradeCancelled) {
        io.emit('updates:upgrade-cancelled', { type: 'snap' });
        resolve({ success: false, error: 'Upgrade cancelled', cancelled: true });
      } else if (code === 0) {
        io.emit('updates:upgrade-complete', { type: 'snap', success: true, duration });
        resolve({ success: true, duration });
      } else {
        // snap refresh returns non-zero if no updates, check output
        if (stdout.includes('All snaps up to date') || stderr.includes('All snaps up to date')) {
          io.emit('updates:upgrade-complete', { type: 'snap', success: true, duration, message: 'All snaps up to date' });
          resolve({ success: true, duration, message: 'All snaps up to date' });
        } else {
          io.emit('updates:upgrade-complete', { type: 'snap', success: false, error: stderr });
          resolve({ success: false, error: `snap refresh failed with code ${code}` });
        }
      }
    });

    snap.on('error', (err) => {
      activeUpgradeProcess = null;
      io.emit('updates:upgrade-complete', { type: 'snap', success: false, error: err.message });
      resolve({ success: false, error: err.message });
    });
  });
}

File: crates/hr-api/src/routes/updates.rs
use axum::{
    extract::State,
    routing::{get, post},
    Json, Router,
};
use hr_common::events::UpdateEvent;
use serde_json::{json, Value};
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::sync::broadcast;
use tracing::error;

use crate::state::ApiState;

static CHECK_RUNNING: std::sync::atomic::AtomicBool = std::sync::atomic::AtomicBool::new(false);
static UPGRADE_RUNNING: std::sync::atomic::AtomicBool =
    std::sync::atomic::AtomicBool::new(false);

pub fn router() -> Router<ApiState> {
    Router::new()
        .route("/status", get(check_status))
        .route("/last", get(last_check))
        .route("/check", post(run_check))
        .route("/cancel", post(cancel_check))
        .route("/upgrade/status", get(upgrade_status))
        .route("/upgrade/apt", post(upgrade_apt))
        .route("/upgrade/apt-full", post(upgrade_apt_full))
        .route("/upgrade/snap", post(upgrade_snap))
        .route("/upgrade/cancel", post(cancel_upgrade))
}

const LAST_CHECK_PATH: &str = "/var/lib/server-dashboard/last-update-check.json";

async fn check_status() -> Json<Value> {
    let running = CHECK_RUNNING.load(std::sync::atomic::Ordering::Relaxed);
    Json(json!({"success": true, "running": running}))
}

async fn last_check() -> Json<Value> {
    match tokio::fs::read_to_string(LAST_CHECK_PATH).await {
        Ok(content) => match serde_json::from_str::<Value>(&content) {
            Ok(result) => Json(json!({"success": true, "result": result})),
            Err(_) => Json(json!({"success": true, "result": null})),
        },
        Err(_) => Json(json!({"success": true, "result": null})),
    }
}

async fn run_check(State(state): State<ApiState>) -> Json<Value> {
    if CHECK_RUNNING.load(std::sync::atomic::Ordering::Relaxed) {
        return Json(json!({"success": false, "error": "Verification deja en cours"}));
    }

    CHECK_RUNNING.store(true, std::sync::atomic::Ordering::Relaxed);
    let tx = state.events.updates.clone();

    // Spawn the check in background so it streams events via WebSocket
    tokio::spawn(async move {
        let _ = tx.send(UpdateEvent::Started);
        let start = std::time::Instant::now();

        // Phase 1: apt update
        let _ = tx.send(UpdateEvent::Phase {
            phase: "apt-update".to_string(),
            message: "Mise a jour des listes de paquets...".to_string(),
        });

        stream_command(&tx, "apt", &["update", "-q"]).await;

        // Phase 2: apt list --upgradable
        let _ = tx.send(UpdateEvent::Phase {
            phase: "apt-check".to_string(),
            message: "Verification des paquets APT...".to_string(),
        });

        let apt_output = tokio::process::Command::new("apt")
            .args(["list", "--upgradable"])
            .output()
            .await;

        let apt_packages = match apt_output {
            Ok(o) => parse_apt_list(&String::from_utf8_lossy(&o.stdout)),
            Err(_) => vec![],
        };

        let security_count = apt_packages
            .iter()
            .filter(|p| {
                p.get("isSecurity")
                    .and_then(|s| s.as_bool())
                    .unwrap_or(false)
            })
            .count();

        let _ = tx.send(UpdateEvent::AptComplete {
            packages: apt_packages.clone(),
            security_count,
        });

        // Phase 3: snap refresh --list
        let _ = tx.send(UpdateEvent::Phase {
            phase: "snap-check".to_string(),
            message: "Verification des snaps...".to_string(),
        });

        let snap_output = tokio::process::Command::new("snap")
            .args(["refresh", "--list"])
            .output()
            .await;

        let snap_packages = match snap_output {
            Ok(o) if o.status.success() => {
                parse_snap_list(&String::from_utf8_lossy(&o.stdout))
            }
            _ => vec![],
        };

        let _ = tx.send(UpdateEvent::SnapComplete {
            snaps: snap_packages.clone(),
        });

        // Phase 4: needrestart
        let _ = tx.send(UpdateEvent::Phase {
            phase: "needrestart".to_string(),
            message: "Verification des services...".to_string(),
        });

        let needrestart = tokio::process::Command::new("needrestart")
            .args(["-b"])
            .output()
            .await;

        let needrestart_info = match needrestart {
            Ok(o) => parse_needrestart(&String::from_utf8_lossy(&o.stdout)),
            Err(_) => json!({"kernelRebootNeeded": false, "services": []}),
        };

        let _ = tx.send(UpdateEvent::NeedrestartComplete(needrestart_info.clone()));

        let duration = start.elapsed().as_millis() as u64;

        let summary = json!({
            "totalUpdates": apt_packages.len() + snap_packages.len(),
            "securityUpdates": security_count,
            "servicesNeedingRestart": needrestart_info.get("services").and_then(|s| s.as_array()).map(|a| a.len()).unwrap_or(0)
        });

        let result = json!({
            "timestamp": chrono::Utc::now().to_rfc3339(),
            "duration": duration,
            "apt": { "packages": apt_packages, "securityCount": security_count },
            "snap": { "packages": snap_packages },
            "needrestart": needrestart_info,
            "summary": summary
        });

        if let Ok(content) = serde_json::to_string_pretty(&result) {
            let _ = tokio::fs::write(LAST_CHECK_PATH, &content).await;
        }

        let _ = tx.send(UpdateEvent::Complete {
            success: true,
            summary,
            duration,
        });

        CHECK_RUNNING.store(false, std::sync::atomic::Ordering::Relaxed);
    });

    Json(json!({"success": true, "message": "Verification lancee"}))
}

async fn cancel_check(State(state): State<ApiState>) -> Json<Value> {
    let _ = tokio::process::Command::new("pkill")
        .args(["-f", "apt update"])
        .output()
        .await;
    CHECK_RUNNING.store(false, std::sync::atomic::Ordering::Relaxed);
    let _ = state.events.updates.send(UpdateEvent::Cancelled);
    Json(json!({"success": true}))
}

async fn upgrade_status() -> Json<Value> {
    let running = UPGRADE_RUNNING.load(std::sync::atomic::Ordering::Relaxed);
    Json(json!({"success": true, "running": running}))
}

async fn upgrade_apt(State(state): State<ApiState>) -> Json<Value> {
    run_upgrade(state, "apt", &["upgrade", "-y"]).await
}

async fn upgrade_apt_full(State(state): State<ApiState>) -> Json<Value> {
    run_upgrade(state, "apt", &["full-upgrade", "-y"]).await
}

async fn upgrade_snap(State(state): State<ApiState>) -> Json<Value> {
    run_upgrade(state, "snap", &["refresh"]).await
}

async fn run_upgrade(state: ApiState, cmd: &str, args: &[&str]) -> Json<Value> {
    if UPGRADE_RUNNING.load(std::sync::atomic::Ordering::Relaxed) {
        return Json(json!({"success": false, "error": "Mise a jour deja en cours"}));
    }

    UPGRADE_RUNNING.store(true, std::sync::atomic::Ordering::Relaxed);
    let tx = state.events.updates.clone();
    let cmd = cmd.to_string();
    let args: Vec<String> = args.iter().map(|s| s.to_string()).collect();

    tokio::spawn(async move {
        let upgrade_type = if cmd == "snap" {
            "snap".to_string()
        } else if args.contains(&"full-upgrade".to_string()) {
            "apt-full".to_string()
        } else {
            "apt".to_string()
        };

        let _ = tx.send(UpdateEvent::UpgradeStarted {
            upgrade_type: upgrade_type.clone(),
        });

        let start = std::time::Instant::now();

        let mut child = match tokio::process::Command::new(&cmd)
            .args(&args)
            .env("DEBIAN_FRONTEND", "noninteractive")
            .stdout(std::process::Stdio::piped())
            .stderr(std::process::Stdio::piped())
            .spawn()
        {
            Ok(c) => c,
            Err(e) => {
                let _ = tx.send(UpdateEvent::UpgradeComplete {
                    upgrade_type,
                    success: false,
                    duration: 0,
                    error: Some(e.to_string()),
                });
                UPGRADE_RUNNING.store(false, std::sync::atomic::Ordering::Relaxed);
                return;
            }
        };

        // Stream stdout
        if let Some(stdout) = child.stdout.take() {
            let tx_c = tx.clone();
            tokio::spawn(async move {
                let reader = BufReader::new(stdout);
                let mut lines = reader.lines();
                while let Ok(Some(line)) = lines.next_line().await {
                    let _ = tx_c.send(UpdateEvent::UpgradeOutput { line });
                }
            });
        }

        // Stream stderr
        if let Some(stderr) = child.stderr.take() {
            let tx_c = tx.clone();
            tokio::spawn(async move {
                let reader = BufReader::new(stderr);
                let mut lines = reader.lines();
                while let Ok(Some(line)) = lines.next_line().await {
                    let _ = tx_c.send(UpdateEvent::UpgradeOutput { line });
                }
            });
        }

        let status = child.wait().await;
        let duration = start.elapsed().as_millis() as u64;
        let success = status.map(|s| s.success()).unwrap_or(false);

        let _ = tx.send(UpdateEvent::UpgradeComplete {
            upgrade_type,
            success,
            duration,
            error: if success {
                None
            } else {
                Some("Upgrade failed".to_string())
            },
        });

        UPGRADE_RUNNING.store(false, std::sync::atomic::Ordering::Relaxed);
    });

    Json(json!({"success": true, "message": "Mise a jour lancee"}))
}

async fn cancel_upgrade(State(state): State<ApiState>) -> Json<Value> {
    let _ = tokio::process::Command::new("pkill")
        .args(["-f", "apt upgrade|apt full-upgrade|snap refresh"])
        .output()
        .await;
    UPGRADE_RUNNING.store(false, std::sync::atomic::Ordering::Relaxed);
    let _ = state.events.updates.send(UpdateEvent::UpgradeCancelled);
    Json(json!({"success": true}))
}

/// Stream command output line by line as UpdateEvent::Output
async fn stream_command(tx: &broadcast::Sender<UpdateEvent>, cmd: &str, args: &[&str]) {
    let mut child = match tokio::process::Command::new(cmd)
        .args(args)
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .spawn()
    {
        Ok(c) => c,
        Err(e) => {
            error!("Failed to spawn {}: {}", cmd, e);
            return;
        }
    };

    if let Some(stdout) = child.stdout.take() {
        let reader = BufReader::new(stdout);
        let mut lines = reader.lines();
        while let Ok(Some(line)) = lines.next_line().await {
            let _ = tx.send(UpdateEvent::Output { line });
        }
    }

    let _ = child.wait().await;
}

fn parse_apt_list(output: &str) -> Vec<Value> {
    output
        .lines()
        .filter(|line| line.contains('/'))
        .filter_map(|line| {
            let parts: Vec<&str> = line.split_whitespace().collect();
            if parts.is_empty() {
                return None;
            }
            let name_source: Vec<&str> = parts[0].split('/').collect();
            let name = name_source.first()?.to_string();
            let source = name_source.get(1).unwrap_or(&"").to_string();
            let new_version = parts.get(1).unwrap_or(&"").to_string();
            let current_version = if line.contains("upgradable from:") {
                line.split("upgradable from: ")
                    .nth(1)
                    .map(|v| v.trim_end_matches(']').to_string())
                    .unwrap_or_default()
            } else {
                String::new()
            };

            let is_security = source.contains("security");

            Some(json!({
                "name": name,
                "currentVersion": current_version,
                "newVersion": new_version,
                "isSecurity": is_security
            }))
        })
        .collect()
}

fn parse_snap_list(output: &str) -> Vec<Value> {
    output
        .lines()
        .skip(1)
        .filter(|line| !line.is_empty())
        .filter_map(|line| {
            let parts: Vec<&str> = line.split_whitespace().collect();
            if parts.len() < 2 {
                return None;
            }
            Some(json!({
                "name": parts[0],
                "newVersion": parts.get(1).unwrap_or(&""),
                "publisher": parts.get(4).unwrap_or(&"")
            }))
        })
        .collect()
}

fn parse_needrestart(output: &str) -> Value {
    let kernel_reboot = output.contains("NEEDRESTART-KSTA: 3");
    let services: Vec<String> = output
        .lines()
        .filter(|l| l.starts_with("NEEDRESTART-SVC:"))
        .filter_map(|l| l.split(':').nth(1).map(|s| s.trim().to_string()))
        .collect();

    json!({
        "kernelRebootNeeded": kernel_reboot,
        "services": services
    })
}
